File: templates/matlab/bootstrap_ci.m
function fig = bootstrap_ci()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(2);
    n = 60; n_boot = 4000;
    data = -1.5*(log(rand(n, 1)) + log(rand(n, 1)));   % Gamma(shape=2, scale=1.5)
    theta_hat = mean(data);
    idx = randi(n, n_boot, n);
    boot = mean(data(idx), 2);
    [lo, hi] = bca_interval(data, boot, theta_hat, 0.05);
    fig = figure; hold on;
    histogram(boot, 40, 'Normalization', 'pdf', 'FaceColor', palette('cat', 1), ...
              'FaceAlpha', 0.75, 'EdgeColor', 'w', 'LineWidth', 0.3);
    h1 = xline(theta_hat, '-', 'Color', palette('cat', 2), 'LineWidth', 1.5);
    h2 = xline(lo, '--', 'Color', [0.5 0.5 0.5], 'LineWidth', 1.1);
    xline(hi, '--', 'Color', [0.5 0.5 0.5], 'LineWidth', 1.1);
    xlabel('bootstrap statistic (mean)'); ylabel('density');
    title('Bootstrap distribution of the mean');
    legend([h1 h2], {sprintf('sample mean = %.2f', theta_hat), ...
           sprintf('BCa 95%% CI [%.2f, %.2f]', lo, hi)}, 'Location', 'northeast');
    grid on;
end

function [lo, hi] = bca_interval(data, boot, theta_hat, alpha)
    z0 = sqrt(2)*erfinv(2*mean(boot < theta_hat) - 1);
    n = numel(data);
    jack = (sum(data) - data) / (n - 1);    % leave-one-out means
    d = mean(jack) - jack;
    a = sum(d.^3) / (6*(sum(d.^2))^1.5);
    z = sqrt(2)*erfinv(2*[alpha/2, 1-alpha/2] - 1);
    p = 0.5*erfc(-(z0 + (z0 + z)./(1 - a*(z0 + z)))/sqrt(2));
    q = quantile_lin(boot, p);
    lo = q(1); hi = q(2);
end

function q = quantile_lin(x, p)
    xs = sort(x(:)).'; n = numel(xs);
    pos = p*(n-1) + 1;
    lo_i = min(max(floor(pos), 1), n);
    hi_i = min(max(ceil(pos), 1), n);
    q = xs(lo_i) + (pos - lo_i).*(xs(hi_i) - xs(lo_i));
end

File: templates/python/bootstrap_ci.py
"""bootstrap_ci: Bootstrap 抽样分布（直方图 + BCa 置信区间 + 原始统计量标注）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from theme import apply_theme
from palette import cycle

def _bca(data, boot, theta_hat, alpha=0.05):
    z0 = stats.norm.ppf((boot < theta_hat).mean())
    n = len(data)
    jack = np.array([np.delete(data, i).mean() for i in range(n)])
    d = jack.mean() - jack
    a = (d**3).sum() / (6.0 * ((d**2).sum())**1.5)
    z = stats.norm.ppf([alpha / 2, 1 - alpha / 2])
    p = stats.norm.cdf(z0 + (z0 + z) / (1 - a * (z0 + z)))
    return np.quantile(boot, p)

def make_figure(data=None, n_boot=4000, title='Bootstrap distribution of the mean'):
    apply_theme()
    rng = np.random.default_rng(2)
    if data is None:
        data = rng.gamma(2.0, 1.5, 60)          # 偏态原始样本
    theta_hat = data.mean()
    idx = rng.integers(0, len(data), (n_boot, len(data)))
    boot = data[idx].mean(axis=1)
    lo, hi = _bca(data, boot, theta_hat)
    fig, ax = plt.subplots()
    ax.hist(boot, bins=40, color=cycle(0), alpha=0.75, edgecolor='w',
            linewidth=0.3, density=True)
    ax.axvline(theta_hat, color=cycle(1), linewidth=1.5,
               label=f'sample mean = {theta_hat:.2f}')
    ax.axvline(lo, color='gray', linestyle='--', linewidth=1.1,
               label=f'BCa 95% CI [{lo:.2f}, {hi:.2f}]')
    ax.axvline(hi, color='gray', linestyle='--', linewidth=1.1)
    ax.set_xlabel('bootstrap statistic (mean)')
    ax.set_ylabel('density')
    ax.set_title(title)
    ax.legend(loc='upper right')
    ax.grid(True, linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)
